File: old/backtracking/generate_binary_strings/generate_binary_strings.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

void backtrack(vector<vector<int>>& result, vector<int>& currentLevel, int n,  int explore) {
    // if our current level is part of solution, add it it our final result 
    if (currentLevel.size() == n) {
        result.push_back(currentLevel);
        return;
    }

    // explore
    explore = 0;
    currentLevel.push_back(explore);
    backtrack(result, currentLevel, n, explore);
    currentLevel.pop_back();

    explore = 1;
    currentLevel.push_back(explore);
    backtrack(result, currentLevel, n, explore);
    currentLevel.pop_back();
}


vector<vector<int>> binaryCombinations(int n) {
    vector<vector<int>> result;
    vector<int> currentLevel;
    
    backtrack(result, currentLevel, n, -1);

    return result;
}

int main() {
    int n;
    cout << "enter n to get all the combinations of n binary digits: ";
    cin >> n;
    vector<vector<int>> result = binaryCombinations(n);
    for (auto& combination : result) {
        for (int bit : combination) {
            cout << bit;
        }
        cout << " ";
    }
    return 0;
}

File: old/backtracking/generate_binary_strings/generate_binary_strings.py
def generate_binary_string(n):
    final_result = []
    current_level = []

    def backtrack(current_level, explore=None):
        # if current level is part of solution
        if len(current_level) == n:
            print(
                f"since {current_level} has reached size {n}, going to add it to {final_result}"
            )
            final_result.append("".join(str(bit) for bit in current_level))
            print(
                f"after adding it to final result, final result looks like this: {final_result}"
            )
            print("\n")
            return

        # explore
        explore = 0
        print(f"exploring: {explore}")
        current_level.append(explore)
        print(f"current_level: {current_level}")
        print(f"going to backtrack: backtrack({current_level}, {explore})")
        backtrack(current_level, explore)
        print(
            f"from {current_level} we are going to pop the last value which is {current_level.pop()}"
        )
        print(f"now our current level looks like this: {current_level}")
        print("\n")

        # explore
        explore = 1
        print(f"exploring: {explore}")
        current_level.append(explore)
        print(f"current_level: {current_level}")
        print(f"going to backtrack: backtrack({current_level}, {explore})")
        backtrack(current_level, explore)
        print(
            f"from {current_level} we are going to pop the last value which is {current_level.pop()}"
        )
        print(f"now our current level looks like this: {current_level}")
        print("\n")

    print(f"current_level: {current_level}")
    print(f"going to backtrack: backtrack({current_level}, None)")
    backtrack(current_level)

    return final_result


def main():
    n = int(input("enter n to generate all combinations of bits of size n: "))
    print(generate_binary_string(n))


main()
